sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala

```diff
@@ -717,65 +719,97 @@ case class GpuGenerateExec(
-      withResource(projectedInput) { projIn =>
-        // 1. compute split indices of input batch
-        val splitIndices = generator.inputSplitIndices(
-          projIn,
-          othersProjectList.length,
-          outer,
-          new RapidsConf(conf).gpuTargetBatchSizeBytes)
-        // 2. split up input batch with indices
-        makeSplitIterator(projIn, splitIndices).map { splitIn =>
-          withResource(splitIn) { splitIn =>
-            // 3. apply generation on each (sub)batch
-            val ret = generator.generate(splitIn, othersProjectList.length, outer)
-            numOutputBatches += 1
-            numOutputRows += ret.numRows()
-            ret
-          }
-        }
-      }
+      val projectedInput = GpuProjectExec.projectAndCloseWithRetrySingleBatch(
+        SpillableColumnarBatch(input, SpillPriorities.ACTIVE_ON_DECK_PRIORITY),
+        othersProjectList ++ genProjectList)
+      getSplits(projectedInput, othersProjectList,
+        new RapidsConf(conf).gpuTargetBatchSizeBytes)
+    }
+    new GpuGenerateIterator(splits, generator, othersProjectList.length, outer,
+      numOutputRows, numOutputBatches, opTime)
+  }
+
+  // Split up the input batch and call generate on each split.
+  private def getSplits(cb: ColumnarBatch,
+      othersProjectList: Seq[GpuExpression],
+      targetSize: Long): Seq[SpillableColumnarBatch] = {
+    withResource(cb) { _ =>
+      // compute split indices of input batch
+      val splitIndices = generator.inputSplitIndices(
+      cb, othersProjectList.length, outer, targetSize)
+      // split up input batch with indices
+      makeSplits(cb, splitIndices)
     }
   }
 
-  private def makeSplitIterator(input: ColumnarBatch,
-      splitIndices: Array[Int]): Iterator[ColumnarBatch] = {
+  // Split a ColumnarBatch according to a list of indices.
+  // Returns a list of spillable batches corresponding to the splits.
+  private def makeSplits(input: ColumnarBatch,
+      splitIndices: Array[Int]): Array[SpillableColumnarBatch] = {
     val schema = GpuColumnVector.extractTypes(input)
 
     if (splitIndices.isEmpty) {
-      withResource(GpuColumnVector.from(input)) { table =>
-        Array(GpuColumnVector.from(table, schema)).iterator
-      }
+      // The caller will close input, so we increment here to offset.
+      Array(SpillableColumnarBatch(GpuColumnVector.incRefCounts(input),
+        SpillPriorities.ACTIVE_ON_DECK_PRIORITY))
     } else {
       val splitInput = withResource(GpuColumnVector.from(input)) { table =>
         table.contiguousSplit(splitIndices: _*)
       }
-      splitInput.zipWithIndex.iterator.map { case (ct, i) =>
+      splitInput.zipWithIndex.safeMap { case (ct, i) =>
         closeOnExcept(splitInput.slice(i + 1, splitInput.length)) { _ =>
           withResource(ct) { ct: ContiguousTable =>
-            GpuColumnVector.from(ct.getTable, schema)
+            SpillableColumnarBatch(ct, schema,
+              SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
```

Comment:
These seem appropriate as: ACTIVE_BATCHING_PRIORITY.

Reply:
Actually, this is where I intended to put that comment above.  Thanks!